Update to Django 5.1

StevenMaude opened the pull request:
Fixes #2115.

This requires a workaround of the coding systems tests that use databases dynamically, due to changes in Django 5.1. We agreed in #2115 that working around the new restrictions was the best thing to do, to allow the upgrade without delay. In future, we may have to revisit this code and associated tests, either for future Django changes or to make maintenance and improvements easier. 

TODO:

- [x] [local testing of the site](https://github.com/opensafely-core/opencodelists/pull/2130#issuecomment-2476300602)
- [x] ~repin Django to 5.1.x~ *not doing this*
- [x] upgrade this branch to Django 5.1.6
- [x] recheck with `django-upgrade` before merge (maybe there are newer migrations?)
- [ ] open an issue to consider refactoring the BNF tests that use a pytest fixture instead of data stored as a file

StevenMaude commented:
The issue here is with coding systems test failures, due to the way the tests use databases.

StevenMaude commented:
This also needs [local testing as previously](https://github.com/opensafely-core/opencodelists/pull/2130#issuecomment-2476300602) to ensure everything is OK with the upgrade.

StevenMaude commented:
Current status of this:

* the working tests do pass, but are a bit of a mess still
* it's probably possible to move the general modified `TestCase` class out of each test module into a higher level, in most cases, and remove that duplication
* the `import_data_path` on the `TestCase` isn't always needed
* there's a bit of reviewing which pytest fixtures are being used (maybe the settings fixture is needed in places)
* from what I've looked at of the `base` tests, there's some other magic going on with fixtures (`bnf_release`) that still means that the specified `db_alias` doesn't get picked up by the test, and we get the same `DatabaseOperationForbidden` error

mikerkelly commented on coding_systems/base/tests/test_import_data_utils.py:
```diff
@@ -34,6 +35,55 @@ def bnf_review_version_with_search(bnf_version_with_search):
+    yield coding_systems_tmp_path
+
+
+class BaseDynamicDatabaseTestCase(TestCase):
```
I know this is draft but the naming clash is a bit unfortunate here as this is referring to `base` tests, whereas this sounds like it means the base class for `DynamicDatabaseTestCase` test cases.

StevenMaude replied:
Yep, good spot. I'll fix it when this PR finally gets tidied up.

mikerkelly commented:
Good progress so far, Steve. I think this is pretty close now!

> fixtures aren't easily available on classes when you're using `unittest.TestCase`

[`@pytest.mark.usefixtures`](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#use-fixtures-in-classes-and-modules-with-usefixtures) may help here? I think that's the intended approach for using fixtures in unittest/Django-style class-based test cases. Instead of the `autouse=True` fixtures in the draft `base` / `test_import_data_utils.py` commit.

> - it's probably possible to move the general modified `TestCase` class out of each test module into a higher level, in most cases, and remove that duplication

Would be very good to do this before we move on, as it's easiest to do now and will be useful if this is modified or extended at all in future. Could be a separate PR or even Issue if the `base` tests can be fixed up.

>- the `import_data_path` on the `TestCase` isn't always needed
>- there's a bit of reviewing which pytest fixtures are being used (maybe the settings fixture is needed in places)

Similar.

> - from what I've looked at of the failing `base` tests, there's some other magic going on with fixtures (`bnf_release` and maybe others) that still means that the specified `db_alias` doesn't get picked up by the test, and we get the same `DatabaseOperationForbidden` error

I think this might be a different issue to the others as this test case doesn't seem to be about dynamically creating a database, so the fix applied doesn't help, so it fails as it was before? (Maybe I missed where the fixtures it uses or function it calls ends up doing that.) I haven't looked in detail, let me know if that might help, or I might if I have time.

StevenMaude commented:
NB: the tests ostensibly pass as of 37d15463ff63a8da8e27d0f93057039bbb16c9a3, but that's because several are commented out, while I got one single coding system base test working.

StevenMaude commented:
There's one test left to fix, then I'll start thinking about how the debris can be tidied up.

StevenMaude commented:
I'm now tidying this up on another branch (`steve/django-5.1-fixup`), and when done will push back to this PR branch.

StevenMaude commented:
Local testing flagged up that logout didn't work in Django 5.1 (until I fixed it). That's another good case for "we should have automated end-to-end tests", as in #2135.

StevenMaude commented on coding_systems/base/tests/helpers.py:
```diff
@@ -0,0 +1,84 @@
+    """
+
+    @property
+    def db_alias(self):
+        # The db_alias that will be added temporarily to the DB.
+        raise NotImplementedError(
+            "This test class requires a database alias to be set."
+        )
```
thought: We currently imply only one `db_alias` can be added. This holds for all but one of the tests (one of the `base` coding system tests, where it's handled as a special case).

StevenMaude commented on coding_systems/ctv3/tests/test_import_data.py:
```diff
@@ -25,92 +27,101 @@ def mock_migrate():
-    assert not (db_dir / "ctv3_v_error_20221001.sqlite3.bu").exists()
+class TestImportData(DynamicDatabaseTestCase):
+    db_alias = "ctv3_v1_20221001"
+    import_data_path = str(MOCK_CTV3_IMPORT_DATA_PATH)
```
thought: Have left these paths converted to strings, as this is what the original code did. I'm not sure if it's really required though, as the tests work without doing so.

StevenMaude replied:
Have removed the `str()` from this and similar.

StevenMaude commented on coding_systems/base/tests/test_import_data_utils.py:
```diff
@@ -9,29 +9,98 @@
-
-    return _csr
+class BaseCodingSystemDynamicDatabaseTestCase(DynamicDatabaseTestCase):
+    @pytest.fixture
+    @staticmethod
+    def bnf_csr():
```
thought: It's nice to put this and the other `@staticmethod` fixture inside the class as that's where the other fixtures are, but it's not strictly necessary.

mikerkelly replied:
My weakly-held opinion is to avoid `staticmethod` use pretty much anywhere. A function that doesn't operate on class or instance data in some way isn't really a method as I understand it. It can instead be hinted as pseudo-private to a module, for example, with a leading underscore in the name.

I'm not sure how much of an advantage grouping the fixtures is. `setup_db` and `cleanup_db` are also associated with this class (only used here)... but I'd rather they stay as functions. Grouping that doesn't have a meaning in code I'd rather we do visually with comments or something.

As a comparative, `staticmethod` isn't used anywhere in opencodelists at the moment, and in job-server it's only used with the github API to avoid mocking in a way I'm also a bit dubious about.

StevenMaude commented on coding_systems/dmd/tests/test_import_data.py:
```diff
@@ -13,129 +15,151 @@
-        VPI.objects.using("dmd_2022-100_20221001").first().vmp
-        == VMP.objects.using("dmd_2022-100_20221001").first()
-    )
+    @pytest.fixture
+    def _get_dmd_version_asthma_medication(self, dmd_version_asthma_medication):
+        self.dmd_version_asthma_medication = dmd_version_asthma_medication
```
thought: As mentioned in some of the commit comments, this is a pattern that's needed to get access to a fixture value from within the class. If there's a better way, it's fine to switch to that.

StevenMaude commented on coding_systems/base/tests/test_import_data_utils.py:
```diff
@@ -9,29 +9,98 @@
+            "bnf_import-data_20221201",
+        ]
+        # This addition needs to be done *before* the regular setUp() runs.
+        self.add_to_databases(*db_alias_additions)
```
thought: This is why fixtures like this have to move inside the class, so they can wrangle the `databases` attribute on the test instance.

mikerkelly replied:
Agreed.

StevenMaude commented on coding_systems/base/tests/helpers.py:
```diff
@@ -0,0 +1,84 @@
+            "This test class requires a database alias to be set."
+        )
+
+    coding_system = None
```
question: Could this be better named? It isn't necessary, but is used to derive what `expected_db_path` is.

StevenMaude replied:
I've the tests that use temporary paths into another class, and renamed `coding_system` to be clearer.

StevenMaude commented on coding_systems/base/tests/helpers.py:
```diff
@@ -0,0 +1,84 @@
+            / f"{self.db_alias}.sqlite3"
+        )
+
+    @pytest.fixture(autouse=True)
+    def _configure_tmp_path(self, request):
+        if self.coding_system is not None:
+            request.getfixturevalue("_get_tmp_path")
```
thought: This is slight magic to make the temporary path fixture an optional one. It may not be useful or necessary. It dynamically loads the `_get_tmp_path` fixture for tests that require it.

Jongmassey replied:
What is the nature of the difference between those that do and don't want that behaviour?

StevenMaude commented on coding_systems/bnf/tests/test_import_data.py:
```diff
@@ -162,79 +158,93 @@ def test_import_data_too_many_csv_files(tmp_path):
-    settings, coding_systems_database_tmp_dir, mock_bnf_import_data_path
-):
-    cs_release_count = CodingSystemRelease.objects.count()
+class BNFDynamicDatabaseTestCase(DynamicDatabaseTestCase):
+    # The BNF tests in this module are a special case where they use fixtures
+    # to write temporary files, instead of having the fixtures as repository
+    # files.
+    # This is set to autouse as all the tests using this class,
+    # use import data from a fixture.
+    # It could be generalised by using the "request" fixture and
+    # request.getfixturevalue(), but we don't need to right now.
+    @pytest.fixture(autouse=True)
+    @pytest.mark.usefixtures("mock_bnf_import_data_path")
+    def _set_import_data_from_fixture(self, mock_bnf_import_data_path):
+        self.import_data_path = mock_bnf_import_data_path
```
question: We could simplify this by writing the import data out to a file, as it should then be usable like the other import data for other tests.

The data is used in `import_data()` and, in most cases, exists in the repository as files; it's possible that the reason it wasn't done here, is because a later author chose not to use the same style.

Is this worth doing, or just leave it as is? It's not a terribly complex special case.

Jongmassey replied:
If it's a stylistic choice then making them consistent is a nice opportunity to reduce complexity (although adding more work to this PR). Could refactoring this be deferred to its own issue?

StevenMaude replied:
> Could refactoring this be deferred to its own issue?

Yes, I think that would be fine.

There is an opportunity to get rid of this extra setup. But the extra setup is not too complex: there is an existing pytest fixture that creates the data, and that gets assigned to an attribute on the specific test class. All this happens in the same Python module.

It's just one of the slight inconsistencies I found when trying to generalise the tests using dynamically created databases.

StevenMaude commented on coding_systems/base/tests/helpers.py:
question: Could this file be better named?

mikerkelly replied:
`dynamic_db_helper`?

StevenMaude commented on requirements.prod.in:
```diff
@@ -8,7 +8,7 @@ beautifulsoup4
 crispy-bootstrap4
 dj-database-url
-django~=4.2
+django
```
thought: This should be:

```suggestion
django~=5.1
```

so that we stick to the minor version we've upgraded to, but get security fixes, and can decide when we want to upgrade to Django 5.2.

mikerkelly replied:
I suppose the team maintaining the repo will generally be aware of Django releases and best positioned to decide when the work should scheduled, if we assume that they tend to be difficult to upgrade to (eg due to incompatibility problems). Conversely, keeping this as it is means the automatic updates will ensure these updates reach us, and some may be automatically integrated without issue.

I think I'm neutral on this.

Jongmassey replied:
My instinct is to go with 5.1 given the key changes that a) introduced features and b) broke our current test setup were introduced in 5.1 (AIUI). I guess in that case if similar things would happen 5.1-5.2, the dependa[bot|becky] PR would have failed tests and we'd know that the upgrade would be nontrivial and could pin in then/do the work.

StevenMaude replied:
I asked Lucy about whether pinning Django made sense, and Lucy's reply was that we leave it unpinned, so let's do that.

mikerkelly commented on coding_systems/base/tests/test_import_data_utils.py:
```diff
@@ -9,29 +9,98 @@
+        self.bnf_version_asthma = bnf_version_asthma
 
 
 def setup_db(csr, exclude_last_concept=False):
```
I suppose as this is only for this module it could be `_setup_db` and `_cleanup_db` to make that slightly clearer. (I know this is how these tests were before this PR.)

mikerkelly reviewed (commented):
Thanks Steve, I think this is looking close to done now. I've responded to your comments and raised some of my own but all fairly minor points I think. Some of them also subject to taste, feel free to push back.

I'm only commenting for now as I know Jon wanted to review also, and it would be good to get eyes other than mine as reviewer, as I proposed the general direction and some of this  is based on prototype code I wrote.

StevenMaude commented on coding_systems/dmd/tests/test_import_data.py:
```diff
@@ -13,129 +15,151 @@
+            self.import_data_path,
+            release_name="2022 1.0.0",
             valid_from=date(2022, 10, 1),
         )
```
I think the diff is unhelpfully confusing here; the release version didn't actually change!

(I did try breaking up the changes in this commit, but it didn't seem to help here.)

* The [`1.1.0` version is from `test_import_data_unexpected_file`](https://github.com/opensafely-core/opencodelists/blob/949c3322e245e51537446a112aaa5201efa9a1e1/coding_systems/dmd/tests/test_import_data.py#L99-L111)
* The [`1.0.0` version is from `test_import_data`](https://github.com/opensafely-core/opencodelists/blob/949c3322e245e51537446a112aaa5201efa9a1e1/coding_systems/dmd/tests/test_import_data.py#L29-L54)

StevenMaude commented on coding_systems/dmd/tests/test_import_data.py:
```diff
@@ -13,129 +15,151 @@
+        ):
             import_data(
-                str(MOCK_DMD_IMPORT_DATA_PATH),
-                release_name="2022 1.0.1",
-                valid_from=date(2022, 9, 1),
+                self.import_data_path,
+                release_name="2022 1.1.0",
+                valid_from=date(2022, 10, 1),
             )
```
Answered in [another comment](https://github.com/opensafely-core/opencodelists/pull/2264/files/c8dc264abc74afc7a82d3eadc0c2388fdc8c2e14..949c3322e245e51537446a112aaa5201efa9a1e1#r1941550334).

Jongmassey commented on coding_systems/base/tests/helpers.py:
```diff
@@ -0,0 +1,84 @@
+    """A special TestCase subclass to facilitate use of
+    a dynamically created database.
+
+    db_alias: name of the database that needs to be accessed, str (required)
+    coding_system: name of the coding system, for generating a temporary
+    database directory, str or None
+    import_data_path: name of the path to import data used, str or None
```
+1

StevenMaude commented:
Thanks to Mike and Jon for the helpful review. Based on their comments, I've done the following:

* added another subclass of the `DynamicDatabaseTestCase` to cover the cases using the temporary path
* moved some code comments into docstrings, and expanded on docstrings/comments
* renamed some of the methods/attributes to be clearer
* moved out the two `staticmethod` methods back to module-level
* labelled some module/class-only functions/methods with a single leading underscore to indicate they are internal use 
* renamed the module containing the class from `helpers.py` to `dynamic_db_classes.py`

I've also:

* upgraded again to Django 5.1.6 which was released 2025-02-06, and contains a couple of bug fixes
* ran django-upgrade again, with no changes

The new commits are those following 4b400217002487253eff2366fd5b75ba8f9c522b. Earlier commit IDs have changed because I've rebased on newer `main` a couple of times.

I left these post-review fix commits deliberately separate, and small for easier review.

When the review is approved, I could squash those commits back to where the relevant code is changed in each case. The previous commits reviewed typically made all the changes to a single file in a single commit (which makes logical sense as each Python module covers one aspect of the tests). That might also make sense once we've finalised these changes.

mikerkelly commented on coding_systems/base/tests/helpers.py:
```diff
@@ -36,10 +36,23 @@ def db_alias(self):
+        are creating during, and accessed in, tests.
+
+        We can't patch this directly in the test case as the class is
+        constructed dynamically.
```
I think it's not completely clear what "the class" means here or what specifically is doing the constructing. "Pytest constructs its own dynamic test class based on the test case class defined in the test module rather than instantiating an instance of it directly"?

mikerkelly reviewed (approved):
Thanks Steve. I think everything looks good now (one minor docs comment). Phew, good work! I think this is deployable, so approving.

It's up to you whether to rebase further or not.

StevenMaude commented:
I've squashed the subsequent post-review commits back in to where they would have gone, if I'd got it looking good the first time.
